homeassistant/components/heos/__init__.py

```diff
@@ -88,7 +103,104 @@
 async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
     """Unload a config entry."""
     controller = hass.data[DOMAIN][DATA_CONTROLLER]
+    controller.dispatcher.disconnect_all()
     await controller.disconnect()
     hass.data.pop(DOMAIN)
     return await hass.config_entries.async_forward_entry_unload(
         entry, MEDIA_PLAYER_DOMAIN)
+
+
+class SourceManager:
+    """Class that manages sources for players."""
+
+    def __init__(self, favorites, inputs, *,
+                 retry_delay: int = COMMAND_RETRY_DELAY,
+                 max_retry_attempts: int = COMMAND_RETRY_ATTEMPTS):
+        """Init input manager."""
+        self.retry_delay = retry_delay
+        self.max_retry_attempts = max_retry_attempts
+        self.favorites = favorites
+        self.inputs = inputs
+        self.source_list = self._build_source_list()
+
+    def _build_source_list(self):
+        """Build a single list of inputs from various types."""
+        source_list = []
+        source_list.extend([favorite.name for favorite
+                            in self.favorites.values()])
+        source_list.extend([source.name for source in self.inputs])
+        return source_list
+
+    async def play_source(self, source: str, player):
+        """Determine type of source and play it."""
+        index = next((index for index, favorite in self.favorites.items()
+                      if favorite.name == source), None)
+        if index is not None:
+            return await player.play_favorite(index)
```

Comment:
What are we returning here? We don't seem to use the return value.

Reply:
They return a bool indicating if the receiver accepted the command, but as you noticed we’re not using it downstream.  I was going to update the lib to raise an exception for all the player commands that fail and update the code here accordingly in a future PR.  I can move the return to its own line here meanwhile if you think that’s necessary in the short term.

Reply:
Updating the docstring describing the return value is enough I think.

Reply:
Yeah, though if I'm going to push a new commit and go through build again, might as well move it the direction I ultimately see it going (not returning anything).  :)